---
title: "WGCNA of Pc timecourse"
author: "BSchroeter"
date: "2023-03-20"
output:
  html_document:
    toc: yes
    df_print: paged
  prettydoc::html_pretty:
    theme: hpstr
    highlight: github
  html_notebook:
    theme: lumen
    toc: yes
    toc_float: yes
---

# Purpose

1. Build a weighted and directed network of genes based on the Dual RNA seq data generated from the timecourse of Phytophthora cinnamomi and Lupinus angustifolius 
2. Determine modules (clusters of genes) from the network associated to disease progression (for example, biotrophic or necrotrophic stages) 
3. Extract the names of the genes from these hubs and compare them to the genes in the literature 
4. Functionally annotate these genes

Working idea: This network will then be used to identify hub genes for each of the early, middle and late interactions.

## Load Libraries

```{r message=FALSE, warning=FALSE, include=FALSE, paged.print=TRUE}
suppressPackageStartupMessages({c(
library(WGCNA),
library(DESeq2),
library(GEOquery),
library(tidyverse),
library(gridExtra),
library(reshape2),
library(rtracklayer),
library(dplyr))})
```

## Load Phytophthora cinnamomi gtf file, clean up to get Gene IDs

## Combine alternatively spliced transcripts (seen in transcript_id as -RA and RB)

```{r}
pc.gtf <- rtracklayer::import("https://ftp.ncbi.nlm.nih.gov/genomes/genbank/protozoa/Phytophthora_cinnamomi/latest_assembly_versions/GCA_018691715.1_ASM1869171v1/GCA_018691715.1_ASM1869171v1_genomic.gtf.gz")
pc.df <- as.data.frame((pc.gtf))

pc.genes <- filter(pc.df, type == "start_codon")
pc.geneids <- select(pc.genes, gene_id, product, protein_id)  

pc.annotations <- pc.geneids %>% 
  distinct(gene_id, .keep_all = T)
nrow(pc.annotations)

pcannot <- as.matrix(pc.annotations)

#write.table(pcannot, file = "pcgeneids.tsv",quote=FALSE,sep="\t")
```

## Assemble count matrix and coldata file

```{r, echo=FALSE}
tmp<-read.table("3col.tsv.gz",header=F)
x<-as.matrix(acast(tmp, V2~V1, value.var="V3"))
colnames(x)<-sapply(strsplit(colnames(x),"_"),"[[",1)

#head(x)
#write.table(x,file="countmatrix.tsv",quote=FALSE,sep="\t")
```

## Import files into r

```{r}
coldata <- read.table('sample_info.tsv')
Pcgenenames <- read.table('pcgeneids.tsv', row.names = 1, quote = "", sep='\t', fill = TRUE, header = TRUE)
IUM83Tanjilcountmatrix <- read.table('countmatrix.tsv') 
```

## Clean countmatrix for P. cinnamomi analysis

```{r}
collabels <- colnames(IUM83Tanjilcountmatrix) <- (coldata$Sample)
colnames(IUM83Tanjilcountmatrix) <- collabels

IUM83Tanjilcountmatrix <- tibble::rownames_to_column(IUM83Tanjilcountmatrix, "gene_id")

IUM83CountMatrix <- IUM83Tanjilcountmatrix %>% 
  filter(str_detect(gene_id, "IUM83")) 
  
rownames(IUM83CountMatrix) <- IUM83CountMatrix$gene_id
IUM83CountMatrix <- IUM83CountMatrix [ ,-1]
IUM83CountMatrix <- as.data.frame(IUM83CountMatrix)
IUM83CountMatrix <- IUM83CountMatrix[ ,-(1:12)]

#write.table(IUM83CountMatrix, file = "Pc_countmatrix.tsv",quote=FALSE,sep="\t")
```

## Clean coldata for Deseq2 normalisation

```{r}
colData <- coldata %>% 
  filter(str_detect(Sample, "Pc"))

rownames(colData) <- colData$Sample
colData <- colData [ ,-1] 
```

## Detect and remove outlier genes

```{r}
#Call a function from WGCNA package that detects outliers 
#Rows should be the samples and the columns genes
gsg <- goodSamplesGenes(t(IUM83CountMatrix))
summary(gsg)

#If false, then there are outliers
gsg$allOK #False

table(gsg$goodGenes) #3012to be excluded
table(gsg$goodSamples) #All 12 samples passed

# remove genes that are detectd as outliers
data <- IUM83CountMatrix[gsg$goodGenes == TRUE,]
```

## Detect outlier samples

```{r}
# detect outlier samples - hierarchical clustering
htree <- hclust(dist(t(data)), method = "average")
groups <- cutree(htree, k=2) # cut tree into clusters

plot(htree, labels(groups))
# draw dendogram with red borders around the clusters
rect.hclust(htree, k=2, border="red")
```

## PCA showing clusters

```{r}
pca <- prcomp(t(data))
pca.dat <- pca$x

pca.var <- pca$sdev^2
pca.var.percentage <- round(pca.var/sum(pca.var)*100, digits = 2)

pca.dat <- as.data.frame(pca.dat)

ggplot(pca.dat,aes(PC1, PC2)) +
  geom_point() +
  geom_text(label = row.names(pca.dat)) + 
  labs(x = paste0('PC1: ', pca.var.percentage[1], ' %'),
       y = paste0('PC2: ', pca.var.percentage[2], ' %'))
```

# Remove outlier samples

```{r}
clusters <- as.data.frame(groups)
head(clusters)
table(clusters)

# the biggest cluster is cluster 2 and the rest will be tagged as outliers
outliers <- clusters #%>% filter(., !groups == 2) %>% row.names()

# exclude outlier samples
data.subset <- data[,!(colnames(data) %in% outliers)]
```

# Normalisation using DESeq2 package

The WGCNA library requires normalisation using the vst (variance-stabilising transform) function from the DESeq2 package

```{r}
# exclude outlier samples from the column data to match the new data
#colData <- coldata %>% 
  #filter(!row.names(.) %in% outliers)

# making sure the rownames and column names identical for the two data sets
#all(rownames(colData) == colnames(data))

# create DESeq Data Set
dds <- DESeqDataSetFromMatrix(countData = IUM83CountMatrix,
                              colData = colData,
                              design = ~ 1) # no model, for normalisation only

# remove all genes with counts < 10 in more than 90% of samples (12*0.9=10.8 ~ 11) as suggested by WGCNA on RNAseq FAQ (https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html)
# <10 in more than 90% samples
dds75 <- dds[rowSums(counts(dds) >= 15) >= 11,]
nrow(dds75) # 5403 genes

# >= 15) >= 9,] 8084 genes
# >= 10) >= 9,] 8911 genes 
# >= 10) >= 11,] 5403 genes #less than 10 counts in %90 of samples
# >= 15) >= 11,] 4420 genes

gene.list <- row.names(dds75)
#lapply(gene.list, write, "all_genes.txt", append=TRUE)

# perform variance stabilization
dds_norm <- vst(dds75)

# get normalized counts
norm.counts <- assay(dds_norm) %>% 
  t()

#Save normalised dataset for faster analysis next time
saveRDS(norm.counts, "norm.counts.rds")
```

# Network Construction

```{r}
# load norm.counts object
#norm.counts <- readRDS("norm.counts")

# Choose a set of soft-thresholding powers to create a scale-free network
power <- c(c(1:10), seq(from = 12, to = 50, by = 2))

# Call the network topology analysis function
sft <- pickSoftThreshold(norm.counts,
                  powerVector = power,
                  networkType = "signed",
                  verbose = 5)


sft.data <- sft$fitIndices

# Visualise Rsquared and connectivity to pick power value
# Select a power that gives maximum R.squared(SFT.R.sq) and minimum connectivity(mean.k)

a1 <- ggplot(sft.data, aes(Power, SFT.R.sq, label = Power)) +
  geom_point() +
  geom_text(nudge_y = 0.1) +
  geom_hline(yintercept = 0.85, color = 'red') +
  labs(x = 'Power', y = 'Scale free topology model fit, signed R^2') +
  theme_classic()


a2 <- ggplot(sft.data, aes(Power, mean.k., label = Power)) +
  geom_point() +
  geom_text(nudge_y = 0.1) +
  labs(x = 'Power', y = 'Mean Connectivity') +
  theme_classic()
  
grid.arrange(a1, a2, nrow = 2) 

# Soft threshold value according to graphs
#To a soft threshold (power) with the scale free topology criterion:
# aim for reasonably high scale free R^2 (column 2), higher than say .80
# and negative slope (around -1, col 4). 
soft_power <- 20
```

# Create network and identify modules

Note: this chunk is set not to run since the blockwise Modules function takes a long time to run. The R object has been saved and is loaded in the next chunk for a faster run time.

```{r eval=FALSE, include=FALSE}
# convert matrix to numeric
norm.counts[] <- sapply(norm.counts, as.numeric)

#assign a temporary variable to avoid namespace error and force WGCNA to use its own cor function
temp_cor <- cor
cor <- WGCNA::cor

# Parameters to be tweaked later
bwnet <- blockwiseModules(norm.counts,
                 maxBlockSize = 15000, # selected total number of genes since CPU memory is sufficient (32GB workstation should be able to handle perhaps 30000)
                 TOMType = "signed",
                 power = soft_power,
                 mergeCutHeight = 0.25,
                 numericLabels = FALSE, # set as false to assign color as labels
                 randomSeed = 1234, # for reproducibility since this function uses clustering
                 verbose = 3)

# Save bwnet object
saveRDS(bwnet, "bwnet.rds")

cor <- temp_cor
```

# Check quality of modules

plot TOM matrix (see yeast tutorial) \# Check if the network follows a scale-free toplogy plot log10(p(K)) vs log10(k) and measure R\^2 \# Get the Module Eigengenes

```{r}
# load bwnet object
bwnet <- readRDS("bwnet.rds")

module_eigengenes <- bwnet$MEs %>%
  orderMEs(.)

head(module_eigengenes)


# Plot the dendrogram and the module colors before and after merging underneath
plotDendroAndColors(bwnet$dendrograms[[1]], cbind(bwnet$unmergedColors, bwnet$colors),
                    c("unmerged", "merged"),
                    dendroLabels = FALSE,
                    addGuide = TRUE,
                    hang= 0.03,
                    guideHang = 0.05)

 # grey module = all genes that doesn't fall into other modules

table(bwnet$colors)
```

# Relate the modules to stages

```{r}
# create traits file - assign 1 if a sample is a certain stage, else assign 0
Dis_traits <- colData %>% 
  mutate(Dis.vs.all = ifelse(grepl('Treated', Treatment), 1, 0)) %>% 
  select(4)

factor_levels <- unique(colData$Stage)

# transform stages into factors and define levels
colData$Stage <- factor(colData$Stage, levels = factor_levels)

traits <- binarizeCategoricalColumns(colData$Stage,
                           includePairwise = FALSE,
                           includeLevelVsAll = TRUE,
                           minCount = 1)

traits <- cbind(Dis_traits, traits)
```

# Visualise module-trait association as a heatmap

```{r}
# Define numbers of genes and samples
nSamples <- nrow(norm.counts)
nGenes <- ncol(norm.counts)

# correlation for module eigengenes and traits
module.trait.corr <- cor(module_eigengenes, traits, use = 'p')
module.trait.corr.pvals <- corPvalueStudent(module.trait.corr, nSamples)

# Heat map v2 (from WGCNA tutorial)
textMatrix =  paste(signif(module.trait.corr, 2), "\n(",
                    signif(module.trait.corr.pvals, 1), ")", sep = "")
dim(textMatrix) = dim(module.trait.corr)
#par(mar = c(6, 8.5, 3, 3))

labeledHeatmap(Matrix = module.trait.corr,
               xLabels = colnames(module.trait.corr),
               yLabels = rownames(module.trait.corr),
               ySymbols = rownames(module.trait.corr),
               colorLabels = FALSE,
               colors = blueWhiteRed(50),
               textMatrix = textMatrix,
               setStdMargins = FALSE,
               cex.lab.y = 0.6,
               cex.text = 0.7,
               zlim = c(-1,1),
               main = paste("Module-trait relationships"))

# get number of genes for each module
table(bwnet$colors)

#Tag genes with module membership and store it in a table
module.gene.mapping <- as.data.frame(bwnet$colors)

# Looking at Early interaction, the turquoise module is highly significant (0.77) and will be extracted for further analysis 
Early <- module.gene.mapping %>% 
  filter(`bwnet$colors` == 'turquoise') %>% 
  rownames()

## Best guess Yellow gene module is that involved in disease (genes expressed in planta)
## Turquoise is important in early and green in late
```
Significance in brackets shows how significantly associated the module is the trait of interest  



# Gene Significance of Stage TP

```{r}
# Define a gene significance variable for Early
GS.Early <- as.numeric(cor(norm.counts, traits$data.Early.vs.all, use = "p"))

# This translates the numeric values into colors
GS.stage_earlyColor = numbers2colors(GS.Early, signed = T)
blocknumber = 1

moduleLabelsAutomatic = bwnet$colors
# Convert labels to colors for plotting
moduleColorsAutomatic = labels2colors(moduleLabelsAutomatic)

datColors = data.frame(moduleColorsAutomatic, GS.stage_earlyColor)[bwnet$blockGenes[[blocknumber]], 
    ]

# Plot the dendrogram and the module colors underneath
plotDendroAndColors(bwnet$dendrograms[[blocknumber]], colors = datColors, groupLabels = c("Module colors", 
    "GS.stage.Early"), dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05)
```

# Intramodular gene analysis

The module membership/intramodular connectivity is calculated as the correlation between the eigengene and the gene expression profile.

```{r}
datKME <- signedKME(norm.counts, module_eigengenes)

# Measure the correlation of the gene's module membership and the associated p-values
module.membership.measure <- cor(module_eigengenes, norm.counts, use = 'p')
module.membership.measure.pvals <- corPvalueStudent(module.membership.measure, nSamples)

# Calculate the gene significance and associated p-values
gene.signf.corr <- cor(norm.counts, traits$data.Early.vs.all, use = 'p')
gene.signf.corr.pvals <- corPvalueStudent(gene.signf.corr, nSamples)


colorOfColumn = substring(names(datKME), 4)#electModules = c("turquoise", "brown", "black", "yellow")
selectModules = colorOfColumn[colorOfColumn !="grey"]

#par(mfrow = c(3, length(selectModules)/3))
  for (module in selectModules) {
      column = match(module, colorOfColumn)
      restModule = moduleColorsAutomatic == module
      verboseScatterplot(datKME[restModule, column], GS.Early[restModule], xlab = paste("Module Membership", 
          module, "module"), ylab = "GS.stage_Early", main = paste("kME.", module, 
          "vs. GS"), col = module)
  }
```

# Calculate the module membership and the associated p-values

# The module membership/intramodular connectivity is calculated as the correlation of the eigengene and the gene expression profile.

# This quantifies the similarity of all genes on the array to every module.

```{r}
module.membership.measure <- cor(module_eigengenes, norm.counts, use = 'p')
module.membership.measure.pvals <- corPvalueStudent(module.membership.measure, nSamples)


module.membership.measure.pvals[1:6,1:8]


# Calculate the gene significance and associated p-values

gene.signf.corr <- cor(norm.counts, traits$data.Early.vs.all, use = 'p')
gene.signf.corr.pvals <- corPvalueStudent(gene.signf.corr, nSamples)


GSEarly <- gene.signf.corr.pvals %>% 
  as.data.frame() %>% 
  arrange(V1) %>% 
  head(50)


# Using the gene significance you can identify genes that have a high significance for trait of interest 
# Using the module membership measures you can identify genes with high module membership in interesting modules.
```

# Look at significant early genes

```{r}
EarlyGS <- GSEarly
EarlyGS <- tibble::rownames_to_column(EarlyGS, "gene_id")
EarlyGS
```

# Add gene ids

```{r}
EarlyDIS <- dplyr::right_join(Pcgenenames, EarlyGS, by = 'gene_id')
EarlyDIS %>%
  arrange(V1)
```

### Session Information

```{r Session Info, echo=FALSE}
sessionInfo()
```
